### PR TITLE
rework

### PR DESCRIPTION
**Pull request by Salbei2500**

### CODE REVIEW

File: functions/client/fn_camRotate.sqf [hunk 1]
```diff
@@ -1,36 +1,53 @@
...
-    waitUntil {camCommitted _camera};
-
-    _angle = if (_cclockwise) then {_angle - _angleDistance} else {_angle + _angleDistance};
-
-    systemChat format ["angle: %1", _angle];
-};
\ No newline at end of file
+params ["_cam", "_target", "_startingAngle", "_endAngle", "_duration", "_radius", ["_clockwise", true], ["_rise", 0]];
+
+GRAD_introCam_camRotateFinisch = false;
```
Comment (nomisum) [inline functions/client/fn_camRotate.sqf]: finisch? 😄

Comment (Salbei2500) [reply functions/client/fn_camRotate.sqf]: yes, a global var to tell you when the rotation is finisched.
Edit: Ah now I see what you mean, I think I'll just remove that c 🗡

File: functions/client/fn_camRotate.sqf [hunk 2]
```diff
@@ -1,36 +1,53 @@
...
+private _pos = getPos _cam;
+private _camAttachObj = "HeliHEmpty" createVehicleLocal _pos;
+_camAttachObj setPos _pos;
+_cam attachTo [_camAttachObj, [0, 0, 0]];
+
+[
+    {
+        params ["_args", "_handle"];
+        _args params ["_camAttachObj", "_cam", "_target", "_steps", "_endAngle", "_duration", "_riseSteps", "_radius"];
+        if (GRAD_introCam_camAngle == _endAngle || {time > _duration}) exitWith {
```
Comment (Fusselwurm) [inline functions/client/fn_camRotate.sqf]: `_duration` is a misnomer, it should be called `_endTime` or something else indicating a *point* in time.

File: functions/client/fn_camRotate.sqf [hunk 3]
```diff
@@ -1,36 +1,53 @@
...
+        };
+
+        GRAD_introCam_camAngle = GRAD_introCam_camAngle + _steps;
+
+        private _newPos = _target getPos [_radius, GRAD_introCam_camAngle];
+        private _height = (getPosASL _camAttachObj) select 2;
+        if (_riseSteps != 0) then {
+            _newPos set [2, (_height + (_riseSteps))];
+        }else{
+            _newPos set [2, _height];
```
Comment (Fusselwurm) [inline functions/client/fn_camRotate.sqf]: `_riseSteps` should be guaranteed to be a number, and then the `else` branch can be removed.

File: functions/client/fn_camRotate.sqf [hunk 4]
```diff
@@ -1,36 +1,53 @@
...
+private _angleDistance = (_endAngle - _startingAngle) mod 360;
+private _steps = if (_clockwise) then {
+    (_angleDistance / _duration)
+}else{
+    (0-(_angleDistance / _duration))
+};
+
+private _riseSteps = if (_rise != 0) then {
+        (_rise / _duration)
+};
```
Comment (Fusselwurm) [inline functions/client/fn_camRotate.sqf:24]: `else {0};` plz… you dont want a var to be multiple data types without a reaally good reason.

File: functions/client/fn_camRotate.sqf [hunk 5]
```diff
@@ -1,36 +1,53 @@
...
-    _angle = if (_cclockwise) then {_angle - _angleDistance} else {_angle + _angleDistance};
-
-    systemChat format ["angle: %1", _angle];
-};
\ No newline at end of file
+params ["_cam", "_target", "_startingAngle", "_endAngle", "_duration", "_radius", ["_clockwise", true], ["_rise", 0]];
+
+GRAD_introCam_camRotateFinish = false;
+private _angleDistance = (_endAngle - _startingAngle) mod 360;
+private _steps = if (_clockwise) then {
```
Comment (Fusselwurm) [inline functions/client/fn_camRotate.sqf]: :thinking: could be made a bit less complex with 

```sqf
private _steps = _angleDistance / _duration;
if (!_clockwise) then {
    _steps = - _steps;
};
```
( nit picking, plz ignore me)

File: functions/client/fn_camRotate.sqf [hunk 6]
```diff
@@ -1,36 +1,53 @@
...
-    _camera camCommitPrepared 0.1;
-
-    waitUntil {camCommitted _camera};
-
-    _angle = if (_cclockwise) then {_angle - _angleDistance} else {_angle + _angleDistance};
-
-    systemChat format ["angle: %1", _angle];
-};
\ No newline at end of file
+params ["_cam", "_target", "_startingAngle", "_endAngle", "_duration", "_radius", ["_clockwise", true], ["_rise", 0]];
```
Comment (Fusselwurm) [inline functions/client/fn_camRotate.sqf]: one thing: as you are dividing by duration further down, do a check somewhere that `_duration` is > 0 and throw an explicit exception or work around it when its <= 0 
(if this is performance critical, maybe bracketed with a `#ifdef DEBUG_MODE [stuff] #endif` or something similar to be switched on/off)

File: functions/server/fn_shotDefinitions.sqf [hunk 7]
```diff
@@ -12,13 +12,15 @@
...
 private _definitions = [
+	/*
 	["ROTATE", 1, 180, target1, 9, 0.3, 0.1, 150, false],
 	["ATTACHED", cam2, cam3, target1, 3, 0.1, 0.1, [0,0,5]],
 	["FREE", cam4, cam5, target1, 2, 0.1, 0.1, [0,0,10]],
 	["FREE", cam5, cam5, target2, 5, 0.1, 0.3, [0,0,50]],
 	["ATTACHED", cam5, cam5, target2, 3, 0.3, 0.3, [0,0,3]],
 	["ATTACHED", cam5, cam5, target2, 5, 0.3, 0.3, [0,0,10]],
 	["ROTATE", 1, 180, target1, 9, 0.3, 0.1, 150, false]
+	*/
```
Comment (Fusselwurm) [inline functions/server/fn_shotDefinitions.sqf]: so… this is something to be configured by the mission creator? in that case, maybe a global var or mission parameter is in order :thinking:

**Review by Fusselwurm — APPROVED** [review]
:slightly_smiling_face: take my comments as the lazy rambling of someone not having any stakes in it… so feel free to ignore

File: functions/client/fn_camRotate.sqf (hunk 6 above)
Comment (Fusselwurm) [reply functions/client/fn_camRotate.sqf]: ah wait, _duration is for the whole scene, and passed around the functions… ok in that case the check should be in the entry point (whereever that is) :thinking:

File: functions/client/fn_camRotate.sqf (hunk 4 above)
Comment (Fusselwurm) [reply functions/client/fn_camRotate.sqf:24]: nah, actually, if in turn `_rise` was to be guaranteed to be a number (and isnt this the case, with the default value in the _params list?), you need neither `if` nor `else`.

**Comment by nomisum** [comment]
we probably need to rebase to https://github.com/gruppe-adler/Co_MountainLion_III.lythium/blob/master/grad-introCam/ anyway, as this is newer code.

(yes i know this is not the preferred way to work with modules 💃, but i hadnt time to rerelase and update every second when building the mission)

**Comment by Salbei2500** [comment]
wörk wörk :D

**Review by nomisum — APPROVED** [review]
apart from minor formatting/naming inconsistencies in wiki and comments i presume this is working and functionable and can be used in production

**Pull request merged**